### PR TITLE
Fix: Fails in Python Callbacks Abort

### LINKED ISSUE #4375: Python callback hangs WarpX without killing it
I ran into two issues.  First, my simulation paused about 1/4 of the way through and second, it didn't crash.  This means I spent a lot of compute time and still didn't get any of the data I wanted.  Not sure how to fix this.  I tried to use the latest build instructions and the development branch.

I attached my submission script, input script, the output, and the error.
[inputs_PICMI_9_stages.txt](https://github.com/ECP-WarpX/WarpX/files/12912053/inputs_PICMI_9_stages.txt)
[output_16970489.txt](https://github.com/ECP-WarpX/WarpX/files/12912054/output_16970489.txt)
[perlmutter_batch.sbatch.txt](https://github.com/ECP-WarpX/WarpX/files/12912055/perlmutter_batch.sbatch.txt)
[run_milestone.py.txt](https://github.com/ECP-WarpX/WarpX/files/12912060/run_milestone.py.txt)
[WarpX.e16970489.txt](https://github.com/ECP-WarpX/WarpX/files/12912061/WarpX.e16970489.txt)

### PR DESCRIPTION
**Pull request by ax3l**
Make sure that errors in Python callbacks abort the simulation. Before, MPI-parallel simulations would hang.
Fix #4375

Tested with this input file: [PICMI_inputs_3d.py.txt](https://github.com/ECP-WarpX/WarpX/files/12923292/PICMI_inputs_3d.py.txt)
```console
mpirun -n 4 Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py
```
```
STEP 1 starts ...
Python callback 'beforestep' failed!
AttributeError: module 'mpi4py.MPI' has no attribute 'oopsi'

At:
  /home/axel/src/warpx/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py(169): mycallback
  /home/axel/micromamba/envs/warpx-openmp-dev/lib/python3.11/site-packages/pywarpx/callbacks.py(247): callfuncsinlist
  /home/axel/micromamba/envs/warpx-openmp-dev/lib/python3.11/site-packages/pywarpx/callbacks.py(98): __call__
  /home/axel/micromamba/envs/warpx-openmp-dev/lib/python3.11/site-packages/pywarpx/WarpX.py(101): evolve
  /home/axel/micromamba/envs/warpx-openmp-dev/lib/python3.11/site-packages/pywarpx/picmi.py(1930): step
  /home/axel/src/warpx/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py(175): <module>
```
```
# simulation returns
```

### CODE REVIEW

File: Source/Python/WarpX_py.cpp [hunk 1]
```diff
@@ -8,6 +8,9 @@
  */
 #include "WarpX_py.H"
 
+#include <exception>
```
Comment (ax3l) [inline Source/Python/WarpX_py.cpp:12]: For `std::exit`
```suggestion
#include <cstdlib>
#include <exception>
```

File: Source/Python/WarpX_py.cpp [hunk 2]
```diff
@@ -8,6 +8,10 @@
  */
 #include "WarpX_py.H"
 
+#include <cstdlib>
+#include <exception>
```
Comment (ax3l) [inline Source/Python/WarpX_py.cpp:12]: for `std::cerr`:
```suggestion
#include <exception>
#include <iostream>
```

**Review by roelof-groenewald — APPROVED** [review]
Very nice! I'm impressed that the Python error is propagated back to C++. Nicely done pybind11 😄

**Comment by ax3l** [comment]
Supports each direction, pretty nice ☺️

**Pull request merged**